crates/typst/src/layout/inline/shaping.rs

```diff
@@ -267,6 +268,17 @@ impl<'a> ShapedText<'a> {
                     frame.size_mut().x += justification_left.at(self.size)
                         + justification_right.at(self.size);
 
+                    let span = if span_offset == 0 {
+                        shaped.span
+                    } else {
+                        let (span, offset) = shaped.span;
+                        if let Some(offset) = offset.checked_add(span_offset as u16) {
```

Comment:
if we go the checked route, might as well ensure that span_offset doesn't exceed u16